src/lib/Simplify.hs

```diff
@@ -368,8 +363,33 @@ caseComputingEffs scrut alts resultTy = do
   Case scrut alts resultTy <$> foldMapM getEffects alts
 {-# INLINE caseComputingEffs #-}
 
-defuncCase
-  :: Emits o
+defuncCaseCore :: Emits o
+  => Atom CoreIR o -> Type CoreIR o
+  -> (forall o'. (Emits o', DExt o o') => Int -> CAtom o' -> SimplifyM i o' (CAtom o'))
+  -> SimplifyM i o (CAtom o)
+defuncCaseCore scrut resultTy cont = do
+  tryAsDataAtom scrut >>= \case
+    Just (scrutSimp, _) -> do
+      altBinderTys <- caseAltsBinderTys =<< getType scrut
+      defuncCase scrutSimp resultTy \i x -> do
+        let xCoreTy = altBinderTys !! i
+        x' <- liftSimpAtom (Just $ sink xCoreTy) x
+        cont i x'
+    Nothing -> case trySelectBranch scrut of
+      Just (i, arg) -> getDistinct >>= \Distinct -> cont i arg
+      Nothing -> case scrut of
+        SimpInCore (ACase scrutSimp alts _) -> do
+          defuncCase scrutSimp resultTy \i x -> do
+            Abs altb altAtom <- return $ alts !! i
+            altAtom' <- applySubst (altb @> SubstVal x) altAtom
+            cont i altAtom'
+        -- TODO(reviewer): Which cases of Atom can actually appear here, in a
+        -- well-typed source program?  I'd like the error to distinguish between
+        -- "I know this is impossible for reason X" from "This could happen,
+        -- but we just didn't program what to do."
```

Comment:
I think `ACase` might be the only one. It has to be a sum type with at least one non-data case. True CoreIR sum types are represented as either explicit constant constructors, which you handle above, or variables, which we can't have here because we only have SimpIR variables (hmm... maybe top-level CoreIR variables could still exist. I want to remove that possibility separately from this PR so let's ignore that for now). That leaves the embedded `SimpInCore` cases, and of those, `ACase` is the only one that could represent a sum type with non-data cases.

Reply:
OK.